### PR TITLE
Generate selfsigned certificate from configuration

### PR DESCRIPTION
**Pull request by jana-rapava**

### CODE REVIEW

File: src/certgen_certificate_generator_server.cc [hunk 1]
```diff
@@ -27,11 +27,285 @@
 */
 
 #include "fty_certificate_generator_classes.h"
+#include <chrono>
+#include <netdb.h>
+
+// because there is NO std::chrono::days in C++14
+#define HOURS_IN_DAY        24
+
+#define COMMON_NAME         "Web server-ipm2-jana-selfsigned"
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: CN need to be created at runtime => <config name>-<hostname>[-selfsigned]

File: src/certgen_certificate_generator_server.cc [hunk 2]
```diff
@@ -27,11 +27,285 @@
...
+        config.setState(STATE);
+        config.setLocality(LOCALITY);
+        config.setOrganization(ORGANIZATION);
+        config.setOrganizationUnit(ORGANIZATION_UNIT);
+        config.setCommonName(COMMON_NAME);
+        config.setEmail(EMAIL);
+
+        std::list<std::string> ipList = getSystemIPs (AF_INET);
+        ipList.merge (getSystemIPs (AF_INET6));
+        config.setIpList(ipList);
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: IPs need to be set if we are in static mode only. In dhcp we need only FQDN (as the IP will change overtime)

File: src/certgen_certificate_generator_server.cc [hunk 3]
```diff
@@ -27,11 +27,285 @@
...
+
+        config.setDnsList(getSystemDomainNames());
+
+        std::string keyType = certgenConfig.keyConf().keyType();
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: Better to create a function: `static Keys generateKeys(KeyConf & conf)` which could be reuse later

File: src/certgen_certificate_generator_server.cc [hunk 4]
```diff
@@ -27,11 +27,285 @@
...
+                certPem = CertificateX509::selfSignSha256(keyPair, config).getPem();
+            }
+        }
+        else
+        {
+            throw std::runtime_error ("Invalid key type");
+        }
+
+        // TODO: store the certificate in security wallet
+        return certPem;
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: The generation return will return "OK"

File: src/certgen_certificate_generator_server.cc [hunk 5]
```diff
@@ -27,11 +27,285 @@
...
+                Keys keyPair = Keys::generateEC (ECKeyType::PRIME256V1);
+                certPem = CertificateX509::selfSignSha256(keyPair, config).getPem();
+            }
+        }
+        else
+        {
+            throw std::runtime_error ("Invalid key type");
+        }
+
+        // TODO: store the certificate in security wallet
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: You can create a function static `void store(CertificateX509 & cert, StorageConf & conf)` doing nothing for now

File: src/certgen_certificate_generator_server.h [hunk 6]
```diff
@@ -22,8 +22,42 @@
...
+            // List of supported commands with a reference to the handler for this command.
+            std::map<Command, FctCommandHandler> m_supportedCommands;
+
+            // Handlers for all supported commands
+            std::string handleGenerateSelfsignedCertificate(const fty::Payload & params);
+            std::string handleGenerateCSR(const fty::Payload & params);
+            std::string handleImportCertificate(const fty::Payload & params);
+
+            // static helpers
+            std::list<std::string> getSystemIPs (int family);
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.h]: static helpers could be define as static and only in the cpp file

File: src/certgen_certificate_generator_server.cc [hunk 7]
```diff
@@ -27,11 +27,285 @@
...
+        configJson << configFile.rdbuf();
+        configFile.close ();
+
+        cxxtools::SerializationInfo certgenSi;
+        cxxtools::JsonDeserializer deserializer (configJson);
+        deserializer.deserialize (certgenSi);
+
+        certgenSi >>= certgenConfig;
+
+        fty::CertificateConfig config;
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc:207]: create static function for filling out the CertificateConfig

File: src/certgen_certificate_generator_server.cc [hunk 8]
```diff
@@ -27,11 +27,285 @@
...
+            return {"ERROR", e.what()};
+        }
+        catch (...) //show must go one => Log and ignore the unknown error
+        {
+            log_error("Unexpected error: unknown");
+            return {"ERROR", ""};
+        }
+    }
+
+    std::list<std::string> CertificateGeneratorServer::getSystemDomainNames()
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: see https://www.gnu.org/software/libc/manual/html_node/Host-Identification.html to get the fqdn

**Review by perrettecl — CHANGES_REQUESTED** [review]

**Review by stark-dev — APPROVED** [review]

File: src/certgen_certificate_generator_server.cc (hunk 1 above)
Comment (jana-rapava) [reply src/certgen_certificate_generator_server.cc]: Addressed.

File: src/certgen_certificate_generator_server.cc (hunk 2 above)
Comment (jana-rapava) [reply src/certgen_certificate_generator_server.cc]: Addressed.

File: src/certgen_certificate_generator_server.cc (hunk 3 above)
Comment (jana-rapava) [reply src/certgen_certificate_generator_server.cc]: Addressed.

File: src/certgen_certificate_generator_server.cc (hunk 4 above)
Comment (jana-rapava) [reply src/certgen_certificate_generator_server.cc]: Addressed.

File: src/certgen_certificate_generator_server.cc (hunk 5 above)
Comment (jana-rapava) [reply src/certgen_certificate_generator_server.cc]: Addressed.

File: src/certgen_certificate_generator_server.h (hunk 6 above)
Comment (jana-rapava) [reply src/certgen_certificate_generator_server.h]: Addressed.

File: src/certgen_certificate_generator_server.cc (hunk 7 above)
Comment (jana-rapava) [reply src/certgen_certificate_generator_server.cc:207]: Addressed.

File: src/certgen_certificate_generator_server.cc (hunk 8 above)
Comment (jana-rapava) [reply src/certgen_certificate_generator_server.cc]: The described approach reads domain name from /etc/hosts first, where is the hostname withoout domain - so not the one we need.
I used already existing environment variable in the end.

**Pull request merged**